internal/app/secretless/handlers/mysql/handler.go

```diff
@@ -11,48 +11,127 @@ import (
 // These values are pulled from the handler credentials config
-type BackendConfig struct {
+type ConnectionDetails struct {
 	Host     string
 	Port     uint
 	Username string
 	Password string
 	Options  map[string]string
 }
 
-func (bc *BackendConfig) Address() string {
-	return bc.Host + ":" + strconv.FormatUint(uint64(bc.Port), 10)
+// TODO: all these comments will be moved and subsumed by kumbi's work
+// This format of this packet is described here:
+//
+//   https://dev.mysql.com/doc/internals/en/mysql-packet.html
+//
+//  +-------------+----------------+---------------------------------------------+
+//  |    Type     |      Name      |                 Description                 |
+//	+-------------+----------------+---------------------------------------------+
+//  | int<3>      | payload_length | Length of the payload. The number of bytes  |
+//  |             |                | in the packet beyond the initial 4 bytes    |
+//  |             |                | that make up the packet header.             |
+//  | int<1>      | sequence_id    | Sequence ID                                 |
+//  | string<var> | payload        | [len=payload_length] payload of the packet  |
+//  +-------------+----------------+---------------------------------------------+
+//
+type mysqlPacket []byte
+
+func (pkt *mysqlPacket) SequenceId() byte {
+	return (*pkt)[3]
+}
+
+func (pkt *mysqlPacket) SetSequenceId(id byte) {
+	(*pkt)[3] = id
+}
+
+type MySQLConnection struct {
+	conn net.Conn
+	sequenceId byte
+}
+
+
+func NewMySQLClientConnection(conn net.Conn) *MySQLConnection {
+	return &MySQLConnection{sequenceId: 0, conn: conn}
+}
+
+func NewMySQLBackendConnection(conn net.Conn) *MySQLConnection {
+	return &MySQLConnection{sequenceId: 1, conn: conn}
+}
+
+func (c *MySQLConnection) RawConnection() net.Conn {
+	return c.conn
+}
+
+func (c *MySQLConnection) SetConnection(conn net.Conn) {
+	c.conn = conn
+}
+
+func (c *MySQLConnection) write(pkt mysqlPacket) error {
+	pkt.SetSequenceId(c.sequenceId)
+	c.sequenceId++
+	if _, err := protocol.WritePacket(pkt, c.conn); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *MySQLConnection) read() (mysqlPacket, error) {
+	pkt, err := protocol.ReadPacket(c.conn)
+	if err != nil {
+		return nil, err
+	}
+
+	mysqlPkt := mysqlPacket(pkt)
+	c.sequenceId = (&mysqlPkt).SequenceId()
+	c.sequenceId++
+	return pkt, nil
+}
+
+func (cd *ConnectionDetails) Address() string {
+	return cd.Host + ":" + strconv.FormatUint(uint64(cd.Port), 10)
 }
 
 // Handler connects a client to a backend. It uses the handler Config and Providers to
-// establish the BackendConfig, which is used to make the Backend connection. Then the data
+// establish the ConnectionDetails, which is used to make the Backend connection. Then the data
 // is transferred bidirectionally between the Client and Backend.
 //
 // Handler requires "host", "port", "username" and "password" credentials.
+//
 type Handler struct {
```

Comment:
Since this is the main type defined in this file (handler.go) it should be at the top of the doc, with its methods defined immediately underneath.

Other types currently in this file (`ConnectionDetails`, `mysqlPacket`, and `MySQLConnection`) should be defined either in a separate file that is part of the same package or defined below and organized in sections by type in order of decreasing importance, with the type definition at the top and the methods for that type below its definition